src/pns.c

```diff
@@ -0,0 +1,48 @@
+/**
+ * @file pns.c  Push notification service
+ *
+ * Copyright (C) 2023 Cody T.-H. Chiu
+ */
+
+#include <re.h>
+#include <baresip.h>
+#include "core.h"
+
+
+static void destructor(void *data)
+{
+	struct pns *pn = data;
+
+	mem_deref((void*)pn->pn_provider);
+	mem_deref((void*)pn->pn_prid);
+	mem_deref((void*)pn->pn_param);
+}
+
+/**
+ * Allocate a Push Notification Service
+ *
+ * @param pnsp     	Pointer to allocated Push Notification Service
+ * @param provider	Push provider
+ * @param prid    	Push resource ID
+ * @param param 	Push provider optional parameter
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int pns_alloc(struct pns **pnsp, const char *provider,
+		const char *prid, const char *param)
+{
+	struct pns *pns;
+
+	if (!provider || !prid) {
+		warning("pns: provider and prid are required\n");
+		return EINVAL;
+	}
+
+	pns = mem_zalloc(sizeof(*pns), destructor);
+	pns->pn_provider = provider;
```

Comment:
Please use `str_dup` here (and for the other fields) to store a copy of the string.

Reply:
Update: Then you have to `mem_deref` in destructor, too.

Reply:
Done.